homeassistant/components/modbus/modbus.py

```diff
@@ -314,13 +376,120 @@ async def async_setup(self) -> bool:
             self._async_cancel_listener = async_call_later(
                 self.hass, self._config_delay, self.async_end_delay
             )
+        else:
+            self.start_update_listener()
+
         return True
 
+    async def async_create_and_register_enocean_dongle(
+        self, config: dict[str, Any]
+    ) -> None:
+        """Create and register enocean dongle."""
+        from .modbusenoceandongle import ModbusEnOceanDongle
+        from .modbusenoceanwago750adapter import ModbusEnOceanWago750Adapter
+
+        input_address = config[CONF_INPUT_ADDRESS]
+        output_address = config[CONF_OUTPUT_ADDRESS]
+        slave = config[CONF_SLAVE]
+        esp_version = config.get(CONF_ESP_VERSION, 3)
+        # Change as soon as other modbus enocean adapters are supported
+        adapter = ModbusEnOceanWago750Adapter(
+            self, slave, input_address, output_address
+        )
+        dongle = ModbusEnOceanDongle(self.hass, adapter, esp_version)
+        # Register dongle if not another enocean dongle was registered yet
+        if self.hass.config_entries.async_entries(ENOCEAN_DOMAIN):
+            _LOGGER.debug("Register modbus enocean dongle")
+            enocean_data = self.hass.data.setdefault(DATA_ENOCEAN, {})
+            await dongle.async_setup()
+            enocean_data[ENOCEAN_DONGLE] = dongle
+
     @callback
     def async_end_delay(self, args: Any) -> None:
         """End startup delay."""
         self._async_cancel_listener = None
         self._config_delay = 0
+        self.start_update_listener()
+
+    def start_update_listener(self) -> None:
+        """Possibly start monitoring of updates."""
+        for (scan_group, interval_millis) in self._scan_groups.items():
+            _LOGGER.debug(
+                "Register scan listener scan_group=%s, interval_millis=%s",
+                scan_group,
+                interval_millis,
+            )
+            async_track_time_interval(
+                self.hass,
+                self.async_update_function(scan_group),
+                timedelta(milliseconds=interval_millis),
+            )
+
+    def register_update_listener(
+        self,
+        scan_group: str,
+        slave: int,
+        input_type: str,
+        min_address: int,
+        max_address: int,
+        func: Callable[
+            [ModbusResponse | None, int, str, int], Coroutine[Any, Any, None]
+        ],
+    ) -> None:
+        """Register update listener."""
+        _LOGGER.debug(
+            "Register update listener slave=%s, input_type=%s, min_address=%s, max_address=%s in scan_group=%s",
+            slave,
+            input_type,
+            min_address,
+            max_address,
+            scan_group,
+        )
+        update_listeners = self._update_listeners_by_scan_group[scan_group]
```

Comment:
@janiversen Here we register update listeners based on a `scan_group` and `input_type`, regardless of the device itself. This has the advantage, that in the same request different devices can get updated. I created the concept of `scan_group` to be able to query different type of inputs with a different interval. i.e. it makes sense to listen for wall switch changes at a shorter interval than for state changes of lights.